sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1169,3 +1170,34 @@ object ComputeCurrentTime extends Rule[LogicalPlan] {
     }
   }
 }
+
+/**
+ * Replace the `UpCast` expression by `Cast`, and throw exceptions if the cast may truncate.
+ */
+object RemoveUpCast extends Rule[LogicalPlan] {
+  private def fail(from: DataType, to: DataType) = {
+    throw new AnalysisException(
+      s"Cannot up cast ${from.simpleString} to ${to.simpleString} as it may truncate")
```

Comment:
I think it might be good to include the awesome path error messages that we use when we fail to resolve here as well.  We might also suggest ways to fix it (i.e. "either add an explicit cast to the input data or choose a higher precision type in the target object").